siplib: fix build with python >= 3.11

With python 3.11, the PyFrameObject structure members have been
removed from the public C API:

https://docs.python.org/3.11/whatsnew/3.11.html#whatsnew311-c-api-porting
https://docs.python.org/3.11/whatsnew/3.11.html#pyframeobject-3-11-hiding

This patch migrates to the opaque PyFrameObject.

Signed-off-by: Ralf Dragon <hypnotoad@lindra.de>
Upstream: N/A as SIP v4 is no longer supported. 4.19.25 is the last release.

--- host-python-sip-4.19.25.orig/siplib/sip.h
+++ host-python-sip-4.19.25/siplib/sip.h
@@ -1799,7 +1799,7 @@
     int (*api_get_time)(PyObject *, sipTimeDef *);
     PyObject *(*api_from_time)(const sipTimeDef *);
     int (*api_is_user_type)(const sipWrapperType *);
-    struct _frame *(*api_get_frame)(int);
+    PyFrameObject *(*api_get_frame)(int);
     int (*api_check_plugin_for_type)(const sipTypeDef *, const char *);
     PyObject *(*api_unicode_new)(SIP_SSIZE_T, unsigned, int *, void **);
     void (*api_unicode_write)(int, void *, int, unsigned);


--- host-python-sip-4.19.25.orig/siplib/siplib.c
+++ host-python-sip-4.19.25/siplib/siplib.c
@@ -448,7 +448,7 @@
 static int sip_api_get_time(PyObject *obj, sipTimeDef *time);
 static PyObject *sip_api_from_time(const sipTimeDef *time);
 static int sip_api_is_user_type(const sipWrapperType *wt);
-static struct _frame *sip_api_get_frame(int);
+static PyFrameObject *sip_api_get_frame(int);
 static int sip_api_check_plugin_for_type(const sipTypeDef *td,
         const char *name);
 static PyObject *sip_api_unicode_new(SIP_SSIZE_T len, unsigned maxchar,
@@ -13741,13 +13741,13 @@
 /*
  * Return a frame from the execution stack.
  */
-static struct _frame *sip_api_get_frame(int depth)
+static PyFrameObject *sip_api_get_frame(int depth)
 {
-    struct _frame *frame = PyEval_GetFrame();
+    PyFrameObject *frame = PyEval_GetFrame();
 
     while (frame != NULL && depth > 0)
     {
-        frame = frame->f_back;
+        frame = PyFrame_GetBack(frame);
         --depth;
     }
 

